Store catalog and schema names in Function objects for serialization

rustyconover opened the pull request:
Functions created in arbitrary catalogs and schemas require their catalog and schema names to be serialized. Without this, FunctionSerializer fails to locate functions as it only searches SYSTEM_CATALOG and DEFAULT_SCHEMA.

This change adds catalog and schema name storage to Function objects. The corresponding names are set during the construction of ScalarFunctionCatalogEntry, TableFunctionCatalogEntry, and AggregateFunctionCatalogEntry instances, ensuring successful serialization and reference resolution.

This was suggested in discussion on Discord by @Tishj

rustyconover committed:
Store catalog and schema names in Function objects for serialization

Functions created in arbitrary catalogs and schemas require their catalog and schema names to be serialized. Without this, FunctionSerializer fails to locate functions as it only searches SYSTEM_CATALOG and DEFAULT_SCHEMA.

This change adds catalog and schema name storage to Function objects. The corresponding names are set during the construction of ScalarFunctionCatalogEntry, TableFunctionCatalogEntry, and AggregateFunctionCatalogEntry instances, ensuring successful serialization and reference resolution.

This was suggested in discussion on Discord by @Thijs

carlopi commented:
Just a note, that renumbering fields is not OK, I see no CI failed but it's still a no-go.

Basically fields can only added at the end or deleted (leaving a gap that should never be used again).
I would first check if there are higher level suggestions / requests for changes, then there will be ways to adapt this to keep compatibility in most cases while having extra fields (with a default ideally).

Maxxen commented:
Yes

rustyconover committed:
Revise PR to not re-number stored fields.

As noticed by @carlopi and @Maxxen

Instead store the catalog_name and schema_name in the name field and split them out upon deserialization, this is possible because "." is not valid in an identifier name.

Maxxen commented on src/include/duckdb/function/function_serialization.hpp:
```diff
@@ -51,10 +59,29 @@ class FunctionSerializer {
 	static pair<FUNC, bool> DeserializeBase(Deserializer &deserializer, CatalogType catalog_type) {
 		auto &context = deserializer.Get<ClientContext &>();
 		auto name = deserializer.ReadProperty<string>(500, "name");
+		string catalog_name;
+		string schema_name;
+		if (name.find('.') != std::string::npos) {
+			std::vector<std::string> parts;
+			std::stringstream ss(name);
+			std::string token;
+
+			while (std::getline(ss, token, '.')) {
+				parts.push_back(token);
+			}
+
+			if (parts.size() != 3) {
+				throw InternalException(
+				    "DeserializeBase - expected three parts for fully qualified serialized function %s", name);
+			}
+			catalog_name = parts[0];
+			schema_name = parts[1];
+			name = parts[2];
```
I don't think this is the right way to go about this. As far as Im aware functions (like all identifiers) can have dots in their name. This is a macro, but id imagine it would work for functions too.
```sql
CREATE OR REPLACE FUNCTION "foo.bar"() as 1 + 2;
D select "foo.bar"();
┌─────────────┐
│ "foo.bar"() │
│    int32    │
├─────────────┤
│           3 │
└─────────────┘
```

I would instead serialize the catalog and schema as separate properties. Then you can even set the `SYSTEM_CATALOG` and `DEFAULT_SCHEMA` as explicit defaults. This would then cause the catalog and schema fields to be omitted when they are the defaults, ensuring older version of DuckDB could still read the  serialized plans from a later DuckDB if they only contain functions within the default catalog/schema.

Maxxen replied:
I see, that's unfortunate. Although in this case I think it's fine? The `DeserializeFunction` call doesn't actually read anything (it just does a catalog lookup) so I think you could defer doing that until after you've read the new catalog/schema fields (with id 504 and 505) AFTER the 503 "has_serialize". 

Something like this?
```c++
	template <class FUNC, class CATALOG_ENTRY>
	static pair<FUNC, bool> DeserializeBase(Deserializer &deserializer, CatalogType catalog_type) {
		auto &context = deserializer.Get<ClientContext &>();
		auto name = deserializer.ReadProperty<string>(500, "name");
		auto arguments = deserializer.ReadProperty<vector<LogicalType>>(501, "arguments");
		auto original_arguments = deserializer.ReadProperty<vector<LogicalType>>(502, "original_arguments");

		auto has_serialize = deserializer.ReadProperty<bool>(503, "has_serialize");
		auto schema = deserializer.ReadPropertyWithExplicitDefault<string>(504, "function_schema", DEFAULT_SCHEMA);
		auto catalog = deserializer.ReadPropertyWithExplicitDefault<string>(505, "function_catalog", SYSTEM_CATALOG);

		// Now call DeserializeFunction to get the actual function, potentially passing in the catalog and schema?
		auto function = DeserializeFunction<FUNC, CATALOG_ENTRY>(context, catalog_type, name, std::move(arguments),
									 std::move(original_arguments));

		return make_pair(std::move(function), has_serialize);
	}
```

Maxxen replied:
ah no wait, shit, that doesn't work on the serialization side cause you need to write the function data...

Maxxen replied:
So, even though field identifiers are numeric, I don't think they actually have to be "in order", you just can't change the type of existing ones. Meaning, I think you should be able to add new properties with e.g. id `505` and `506` _before_ the existing `504` "function_data". As long as the fields are set to their default values (which they most likely will be for the vast majority of functions) they won't be serialized so you'd preserve forwards-compatibility at least in that case. 

That said, I think we're currently trying to not break forwards compatibility at all for a while, or at least there has been some work in trying to make it explicit/identifiable when a database contains "newer" entities that older versions won't be able to deserialize. But since functions aren't normally serialized as part of the database file anyway Im not sure what the current policy is...

Maxxen replied:
Somewhat related: https://github.com/duckdb/duckdb/pull/15702

rustyconover replied:
@Maxxen do you think we can get this landed in 1.3.x?  Can we break compatibility now?  @carlopi what do you think?

rustyconover commented:
Hi DuckDB team, any chance this is going be resolved in 1.3.x?

rustyconover commented:
Closing due to new PR posted #17512
